### PR TITLE
Avoiding trying to gc images with no tags which are still in use

### PR DESCRIPTION
**Pull request by ronnielai**
#29325

### CODE REVIEW

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -245,6 +245,8 @@ type Container struct {
 	// The image name of the container, this also includes the tag of the image,
 	// the expected form is "NAME:TAG".
 	Image string
+	// The id of the image used by the container,.
```
Comment (Random-Liu) [inline pkg/kubelet/container/runtime.go]: s/,./.

File: pkg/kubelet/image_manager_test.go [hunk 2]
```diff
@@ -75,8 +75,9 @@ func makeImage(id int, size int64) container.Image {
 // Make a container with the specified ID. It will use the image with the same ID.
 func makeContainer(id int) *container.Container {
 	return &container.Container{
-		ID:    container.ContainerID{Type: "test", ID: fmt.Sprintf("container-%d", id)},
-		Image: imageName(id),
+		ID:      container.ContainerID{Type: "test", ID: fmt.Sprintf("container-%d", id)},
+		Image:   imageName(id) + "-name",
+		ImageID: imageName(id),
```
Comment (Random-Liu) [inline pkg/kubelet/image_manager_test.go]: The function `imageName` is a bit misleading here.

File: pkg/kubelet/image_manager_test.go [hunk 3]
```diff
@@ -85,11 +86,21 @@ func TestDetectImagesInitialDetect(t *testing.T) {
...
 	fakeRuntime.AllPodList = []*containertest.FakePod{
 		{Pod: &container.Pod{
 			Containers: []*container.Container{
-				makeContainer(1),
+				{
+					ID:      container.ContainerID{Type: "test", ID: fmt.Sprintf("container-%d", 1)},
+					ImageID: imageName(1),
+					// The image filed is not set to simulate a no-name image
+				},
+				{
```
Comment (Random-Liu) [inline pkg/kubelet/image_manager_test.go]: Why not use make container here?

File: pkg/kubelet/image_manager_test.go [hunk 4]
```diff
@@ -107,6 +118,10 @@ func TestDetectImagesInitialDetect(t *testing.T) {
 	require.True(t, ok)
 	assert.Equal(zero, withContainer.firstDetected)
 	assert.True(withContainer.lastUsed.After(startTime))
+	withContainerUsingNoNameImage, ok := manager.getImageRecord(imageName(2))
```
Comment (Random-Liu) [inline pkg/kubelet/image_manager_test.go]: If I understand correctly, `imageName(1)` corresponds to `withContainerUsingNoNameImage`.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (ronnielai) [inline pkg/kubelet/container/runtime.go]: fixed

File: pkg/kubelet/image_manager_test.go (hunk 2 above)
Comment (ronnielai) [inline pkg/kubelet/image_manager_test.go]: I've changed the name to imageID.

File: pkg/kubelet/image_manager_test.go (hunk 3 above)
Comment (ronnielai) [inline pkg/kubelet/image_manager_test.go]: Just want to make the comparison clear to the reader that one image has no name while the other has one.

File: pkg/kubelet/image_manager_test.go (hunk 4 above)
Comment (ronnielai) [inline pkg/kubelet/image_manager_test.go]: Yeah, the naming is incorrect. Fixed.

**Comment by Random-Liu** [comment]
LGTM. :)

File: pkg/kubelet/image_manager.go [hunk 5]
```diff
@@ -345,10 +345,5 @@ func isImageUsed(image container.Image, imagesInUse sets.String) bool {
 	if _, ok := imagesInUse[image.ID]; ok {
```
Comment (Random-Liu) [inline pkg/kubelet/image_manager.go]: Remove `and all the RepoTags and RepoDigests.`

Comment (ronnielai) [inline pkg/kubelet/image_manager.go]: done

**Pull request merged**